mozjs/js/rust/src/jsslotconversions.rs

```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Conversion of Rust values to and from JS Slots
+
+#![deny(missing_docs)]
+
+use conversions::{ConversionResult, ConversionBehavior, FromJSValConvertible,
+                  ToJSValConvertible};
+use jsapi::root::*;
+use jsval;
+use std::mem;
+
+/// A trait to record the number of slots required for the current type
+pub trait NumSlots {
+    /// Number of slots required to store Self in inheritance
+    const NUM_SLOTS: u32 = 1;
+
+    /// Number of slots required to store Self as pointer
+    const NUM_SLOTS_PTR: u32 = 1;
+}
+
+/// A trait to store and retrieve data with Rust type to JS slots
+pub trait ToFromJsSlots : NumSlots {
```

Comment:
This just made me realize that if we store any types that need destructors to run inside the slots, then we need to implement a finalize hook for the type. We can express that in this trait like this:

```rust
/// Whether this type needs finalization. Don't need finalization by default.
const NEEDS_FINALIZE: bool = false;

/// Only called for types where `NEEDS_FINALIZE` is `true`.
fn finalize(_fop: *mut js::FreeOp, _obj: *mut JSObject, _offset: u32) {
    // Provide a default no-op `finalize` method to make the procedural
    // macro's life easier and not have to figure out how to sort which fields
    // need finalization or not.
}
```

and then the `magic_codegen` can emit the `JSCLASS_FOREGROUND_FINALIZE` flag into the `JSClass` and a finalizer function when

```rust
...
let mut needs_finalize = quote! { false };
...
for ty in field_types {
    ...
    needs_finalize = quote! {
        #needs_finalize || <#ty as ToFromJsSlots>::NEEDS_FINALIZE
    };
    ...
}
```